File: src/cdisc_library_api/specs/sdtmig_dataset_variable.clj
(ns cdisc-library-api.specs.sdtmig-dataset-variable
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [cdisc-library-api.specs.sdtmig-dataset-variable-links :refer :all]
            )
  (:import (java.io File)))


(def sdtmig-dataset-variable-data
  {
   (ds/opt :ordinal) string?
   (ds/opt :name) string?
   (ds/opt :label) string?
   (ds/opt :description) string?
   (ds/opt :role) string?
   (ds/opt :simpleDatatype) string?
   (ds/opt :core) string?
   (ds/opt :describedValueDomain) string?
   (ds/opt :valueList) (s/coll-of string?)
   (ds/opt :_links) sdtmig-dataset-variable-links-spec
   })

(def sdtmig-dataset-variable-spec
  (ds/spec
    {:name ::sdtmig-dataset-variable
     :spec sdtmig-dataset-variable-data}))

File: src/main/scala/org/openapitools/client/api/SdtmigDatasetVariable.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import SdtmigDatasetVariable._

case class SdtmigDatasetVariable (
  ordinal: Option[String],
name: Option[String],
label: Option[String],
description: Option[String],
role: Option[String],
simpleDatatype: Option[String],
core: Option[String],
describedValueDomain: Option[String],
valueList: Option[List[String]],
links: Option[SdtmigDatasetVariableLinks])

object SdtmigDatasetVariable {
  import DateTimeCodecs._

  implicit val SdtmigDatasetVariableCodecJson: CodecJson[SdtmigDatasetVariable] = CodecJson.derive[SdtmigDatasetVariable]
  implicit val SdtmigDatasetVariableDecoder: EntityDecoder[SdtmigDatasetVariable] = jsonOf[SdtmigDatasetVariable]
  implicit val SdtmigDatasetVariableEncoder: EntityEncoder[SdtmigDatasetVariable] = jsonEncoderOf[SdtmigDatasetVariable]
}
